common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,883 @@
+            if (isDuplicateBlock()) {
+              deferredBufs = null;
+              return;
+            }
+            if (partitionInfo.getCurrentMapIndex() < 0) {
+              writeAnyDeferredBufs();
+            }
+            long updatedPos = partitionInfo.getPosition() + length;
+            boolean indexUpdated = false;
+            if (updatedPos - partitionInfo.getLastChunkOffset() >= minChunkSize) {
+              partitionInfo.updateChunkInfo(updatedPos, mapIndex);
+              indexUpdated = true;
+            }
+            partitionInfo.setPosition(updatedPos);
+            partitionInfo.setCurrentMapIndex(-1);
+
+            // update merged results
+            partitionInfo.blockMerged(mapIndex);
+            if (indexUpdated) {
+              partitionInfo.resetChunkTracker();
+            }
+          } else {
+            deferredBufs = null;
+            throw new RuntimeException(String.format("%s %s to merged shuffle",
+              ErrorHandler.BlockPushErrorHandler.BLOCK_APPEND_COLLISION_DETECTED_MSG_PREFIX,
+              msg.blockId));
+          }
+        }
+        isWriting = false;
+      }
+
+      @Override
+      public void onFailure(String streamId, Throwable throwable) throws IOException {
+        if (errorHandler.shouldLogError(throwable)) {
+          logger.error("Encountered issue when merging shuffle partition block {}", msg, throwable);
+        } else {
+          logger.debug("Encountered issue when merging shuffle partition block {}", msg, throwable);
+        }
+        // Only update partitionInfo if the failure corresponds to a valid request. If the
+        // request is too late, i.e. received after shuffle merge finalize, #onFailure will
+        // also be triggered, and we can just ignore. Also, if we couldn't find an opportunity
+        // to write the block data to disk, we should also ignore here.
+        Map<Integer, AppShufflePartitionInfo> shufflePartitions = partitions.get(appShuffleId);
+        if (isWriting && partitionInfo != null && shufflePartitions != null
+          && shufflePartitions.containsKey(reduceId)) {
+          synchronized (partitionInfo) {
+            logger.debug("{} shuffleId {} reduceId {} set encountered failure", appShuffleId.appId,
+              appShuffleId.shuffleId, reduceId);
+            partitionInfo.setCurrentMapIndex(-1);
+            partitionInfo.setEncounteredFailure(true);
+          }
+        }
+      }
+    };
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOException {
+    logger.info("Finalizing shuffle {} from Application {}.", msg.shuffleId, msg.appId);
+    List<RoaringBitmap> bitmaps = new LinkedList<>();
+    List<Integer> reduceIds = new LinkedList<>();
+    List<Long> sizes = new LinkedList<>();
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
```

Comment:
I think converting this map into a 2-level map breaks the atomic guarantee we need.
Inside `onData`, `onComplete`, and `onFailure`, this 2-level map is read.
However, this read is now broken into 2 non-atomic read:
You first performs an atomic read of the first level map, followed by an atomic read of the 2nd level map.
However, together these 2 reads are not atomic.
Here, when you delete the data from this 2 level map, once again it gets broken into 2 separate deletes, which together is not atomic.
This would make things go wrong.

Reply:
Other than the issue pointed below by @Ngone51, do you see any other race conditions?

Reply:
Inside `onFailure`, access to the map should be inside the lock.

Reply:
The concern about race can be addressed by rely on `compute` on the outer map - if we want to ensure atomicity.
It results in widening the lock per application+shuffle though

Reply:
Thanks for pointing it out @Victsm .

I'm wondering that the merging concurrency might be decreased if we use the wider lock?

Reply:
There were some race conditions introduced when changed the map into a 2-level map. However, now all of them are addressed thanks to everyone's input. So, I don't think we need to introduce a wider lock. Please take another look at the current changes.